Simplified Spark implementation, Added Proto support to Spark. (#734)
lightbend/cloudflow: examples/spark-sensors-proto/README.md

|src-33|  gas|1269.5736118755415|
|src-97|  oil|188.75469481408513|
|src-33|  gas|1269.5736118755415|
+------+-----+------------------+
only showing top 20 rows

-------------------------------------------
Batch: 2
-------------------------------------------
+------+-----+------------------+
|   src|gauge|             value|
+------+-----+------------------+
|src-33|  oil| 847.6950249637855|
|src-33|  gas| 894.1615286881954|
|src-33|  oil| 847.6950249637855|
|src-97|  gas| 901.3472093816678|
|src-97|  gas| 1143.234750548319|
|src-97|  gas| 901.3472093816678|
|src-97|  gas| 1143.234750548319|
|src-72|  oil|1571.1498223984206|
|src-21|  oil|148.37603003222227|
|src-72|  oil|1571.1498223984206|
|src-21|  gas|453.74071025583896|
|src-21|  gas|453.74071025583896|
|src-21|  oil|148.37603003222227|
|src-10|  gas|1649.6090844251842|
|src-10|  gas| 178.6005146200035|
|src-92|  oil| 376.5333753643078|
|src-92|  oil| 376.5333753643078|
|src-10|  gas|1649.6090844251842|
|src-10|  gas| 178.6005146200035|
|src-19|  gas|1547.4380898089883|
+------+-----+------------------+
only showing top 20 rows
```

* Undeploy.

```bash
$ kubectl cloudflow undeploy spark-sensors
```
### Protobuf vs Avro

> This project demonstarte usage of protobuf's for message encoding. Protobufs are enabled through usage of

Reviewer: ```suggestion
This project demonstrates the usage of Protobuf for message encoding. Protobuf is enabled through usage of
```

## Spark-based Cloudflow Application using protobuf encoding

### Problem Definition

> In this application we use a an ingress to generate random data from a set of virtual sensors, use Spark to emit moving average values for each sensor id and the report the values to the console egress.

Reviewer: ```suggestion
In this application we use an ingress to generate random data from a set of virtual sensors, use Spark to emit moving average values for each sensor id and the report the values to the console egress.
```

Author: Fixed

All Protobuf message definitions can be found [here](src/main/protobuf).

> Protobuf encoding is enabled through usage of `ProtoInlet` and `Protoutlet` and usage of the `scalapb.spark.Implicits._` import

Reviewer: ```suggestion
Protobuf encoding is enabled through usage of `ProtoInlet` and `Protoutlet` and usage of the `scalapb.spark.Implicits._` import.
```

Author: fixed

Note: `scalapb.spark.Implicits._` import is incompatible with `cloudflow.spark.sql.SQLImplicits._`.

> The two can be used together in the same class. As a result, if an implementation is using any of the custom

Reviewer: ```suggestion
The two cannot be used together in the same class. As a result, if an implementation is using any of the custom
```

Author: Fixed

> scala case classes (or Avro) encoders for these classes have to be defined. For example a

Reviewer: ```suggestion
scala case classes (or Avro), encoders for these classes have to be defined. For example a
```

Author: fixed

|src-33|  oil| 847.6950249637855|
|src-33|  gas| 894.1615286881954|
|src-33|  oil| 847.6950249637855|
|src-97|  gas| 901.3472093816678|
|src-97|  gas| 1143.234750548319|
|src-97|  gas| 901.3472093816678|
|src-97|  gas| 1143.234750548319|
|src-72|  oil|1571.1498223984206|
|src-21|  oil|148.37603003222227|
|src-72|  oil|1571.1498223984206|
|src-21|  gas|453.74071025583896|
|src-21|  gas|453.74071025583896|
|src-21|  oil|148.37603003222227|
|src-10|  gas|1649.6090844251842|
|src-10|  gas| 178.6005146200035|
|src-92|  oil| 376.5333753643078|
|src-92|  oil| 376.5333753643078|
|src-10|  gas|1649.6090844251842|
|src-10|  gas| 178.6005146200035|
|src-19|  gas|1547.4380898089883|
+------+-----+------------------+
only showing top 20 rows
```

* Undeploy.

```bash
$ kubectl cloudflow undeploy spark-sensors
```
### Protobuf vs Avro

This project demonstrates the usage of Protobuf for message encoding. Protobuf is enabled through usage of
[Scala-PB Spark](https://scalapb.github.io/docs/sparksql/).
ScalaPB comes with its own `implicits` - `scalapb.spark.Implicits._` which is not compatible with `cloudflow.spark.sql.SQLImplicits._`
The following are the rules for using imports.

> * When protobufs are not used, you should use `cloudflow.spark.sql.SQLImplicits._` in all classes as required - see example

Reviewer: see example?

Author: fixed

* When protobufs are used, you should use `scalapb.spark.Implicits._`. Here are the caveats. 
You may want to import `cloudflow.spark.sql.SQLImplicits.StringToColumn` to convert $"col name" into a Column.

> If you use custom scala classes in you Spark calculations, you need to explicitely create encoder for it. See

Reviewer: ```suggestion
If you use custom scala classes in you Spark calculations, you need to explicitly create encoder for it. See
```

Author: Fixed

> [SparkRandomGenDataIngress](src/main/scala/sensors/proto/SparkRandomGenDataIngress.scala) for example

Reviewer: ```suggestion
[SparkRandomGenDataIngress](src/main/scala/sensors/proto/SparkRandomGenDataIngress.scala) for example.
```

Author: Fixed

> * Usage of both imports in the same class will lead to a problem

Reviewer: ```suggestion
* Usage of both imports in the same class will lead to problems.
```

Author: Fixed

> Also note dependency overwrites necessary to use the right version of `com.fasterxml.jackson.core` versions

Reviewer: Why is this necessary?

Author: Was there before. Removed, based on changes to cloudflow here
